src/Nethermind/Nethermind.Evm/EvmObjectFormat.cs

```diff
@@ -0,0 +1,206 @@
+                            {
+                                header = null; return false;
+                            }
+
+                            i += 2;
+                            break;
+                        }
+                    default: // if section kind is anything beside a section-limiter or a terminator byte we return false
+                        {
+                            header = null; return false;
+                        }
+                }
+            }
+
+            var contractBody = code[i..];
+            var calculatedCodeLen = (int)header.CodeSize + (int)header.DataSize;
+            if ((header.CodeSize == 0) || (calculatedCodeLen != contractBody.Length))
+            {
+                header = null; return false;
+            }
+
+            return true;
+        }
+
+        public bool ValidateEofCode(Span<byte> code) => ExtractHeader(code, out _);
+        public bool ValidateEofCode(byte[] code) => ExtractHeader(code, out _);
+
+        public (int StartOffset, int EndOffset)? ExtractCodeOffsets(byte[] code)
+        {
+            if(ExtractHeader(code, out var header))
+            {
+                return ExtractCodeOffsets(header);
+            }
+            return null;
+        }
+
+        public int codeStartOffset(EofHeader header) => header.DataSize == 0
+                                                            ? 5 + EofMagicLength  // magic (2b) + version(1b) +  1 * (sectionId(1b) + sectionSize(2b)) + separator(1b) = magic (2b) + 5b
+                                                            : 8 + EofMagicLength; // magic (2b) + version(1b) +  2 * (sectionId(1b) + sectionSize(2b)) + separator(1b) = magic (2b) + 8b
+        public int codeEndOffset(EofHeader header)  => codeStartOffset(header) + header.CodeSize;
+
+        public (int StartOffset, int EndOffset) ExtractCodeOffsets(EofHeader header)
+        {
+            var endIndex = codeEndOffset(header);
+            var strIndex = endIndex - header.CodeSize;
+            return (strIndex, endIndex);
+        }
+        public bool ValidateInstructions(Span<byte> code, out EofHeader header)
+        {
+            // check if code is EOF compliant
+            if (ExtractHeader(code, out header))
+            {
+                var (startOffset, endOffset) = ExtractCodeOffsets(header);
+                Instruction? opcode = null;
+                for (int i = startOffset; i < endOffset;)
+                {
+                    opcode = (Instruction)code[i];
+
+                    // validate opcode
+                    if (!Enum.IsDefined(typeof(Instruction), code[i]))
+                    {
+                        return false;
+                    }
+
+                    if (opcode is >= Instruction.PUSH1 and <= Instruction.PUSH32)
+                    {
+                        i += code[i] - (int)Instruction.PUSH1 + 1;
+                    }
+                    i++;
```

Comment:
before this just add the truncated immediate check by checking if (i >= endOffset) return false;

Reply:
I don't think it is necessary, since if `i += code[i] - (int)Instruction.PUSH1 + 1;` pushes `i` outside the code section boundaries the loop will end leaving `PUSHx` as the last instruction which will return false.
also the code in IL level will have almost 3~4 additional IL instructions that will run every time we encounter a `PUSH` instruction (I added the check but I kind of don't like it being there)